cipher/ecc.c

```diff
@@ -1345,6 +1346,29 @@ ecc_encrypt_raw (gcry_sexp_t *r_ciph, gcry_sexp_t s_data, gcry_sexp_t keyparms)
         mpi_clear_bit (data, i);
       mpi_set_highbit (data, mpi_get_nbits (pk.E.p) - 1);
     }
+
+  /* For GOST extract the UKM value. If its length is unspecified
+     take 64 bits as default. */
+  if ((flags & PUBKEY_FLAG_GOST))
+    {
+      unsigned int ukm_blen = ctx.saltlen ? ctx.saltlen : 64;
+      if (_gcry_mpi_get_nbits (data) < ukm_blen)
+        {
+          rc = GPG_ERR_TOO_SHORT;
+          goto leave;
+        }
+      salt = _gcry_mpi_copy (data);
+      if (!salt)
+        {
+          rc = gpg_error_from_syserror ();
+          goto leave;
+        }
+      _gcry_mpi_clear_highbit (salt, ukm_blen);
+      if (DBG_CIPHER)
+        log_printmpi ("UKM: ", salt);
+      _gcry_mpi_rshift (data, data, ukm_blen);
+    }
```

Comment:
This UKM format has to be documented somewhere.

Reply:
Also `if salt == 0 then salt = 1` should be implemented.

Reply:
* Ok, I'll add some info to the top encrypt/decrypt comment.

> Also `if salt == 0 then salt = 1` should be implemented.

Why?

Reply:
At least you should check that `salt != 0` (see RFC 7836).

Reply:
Added `&& gcry_mpi_cmp_ui (salt, 0)`.